### ISSUE TITLE
What does this sentence mean cone = torch.cat([intrinsic, coords3d[..., -3:], coords3d[..., -90:-87]], dim=-1)？

### ISSUE THREAD

**Issue by shaoqb** [exiawsh/StreamPETR#4]
https://github.com/exiawsh/StreamPETR/blob/main/projects/mmdet3d_plugin/models/dense_heads/streampetr_head.py#L416

coords3d[..., -3:], coords3d[..., -90:-87], what do they represent?

**Comment by exiawsh** [comment]
Hi! The purpose of this operation is to encode the the camera rays (extrinsics) into context. We assume that sampling two points is able to define a camera ray, and intuitively choose two l2_norm friendly points.(I haven't tried any other points, just intuitively)

**Comment by exiawsh** [comment]
This operation is not critical and has little impact on high baseline model…

**Comment by exiawsh** [comment]
In my experience, 1mAP at most. When I was training Focal-PETR earlier, spatial alignment always gave better results, so StreamPETR follows this design. If you want to migrate StreamPETR to another sparse query framework and use streaming training, you'd better to implement query denosing, recently I found it's necessaey for training stability…（sliding window training dosen't have this problem, however sliding window cosumes 4-6x training time）

**Comment by shaoqb** [comment]
How much impact does the FocalHead have in training Streampetr？

**Comment by exiawsh** [comment]
1 mAP on V2-99 Backbone，but I did not do ablation on R50 Backbone

**Comment by shaoqb** [comment]
https://github.com/exiawsh/StreamPETR/blob/main/projects/mmdet3d_plugin/datasets/nuscenes_dataset.py#L176
Why here is ego_pose = e2g_matrix@l2e_matrix instead of ego_pose = l2e_matrix@e2g_matrix？Isn't egopose the conversion matrix from lidar to global?

**Comment by exiawsh** [comment]
I'm sure that the left multiplication of the matrix is correct. Because the position will be right multiply. pos_xyz_lidar → pos_xyz_ego→ pos_xyz_global

**Comment by shaoqb** [comment]
Yes, you are right
